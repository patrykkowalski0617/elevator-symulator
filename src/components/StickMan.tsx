import React from "react";
import styled from "styled-components";

const StickManStyled = styled.div`
    background-color: #222;
    width: 5px;
    height: 25px;
    position: absolute;
    bottom: 0;
    &::before {
        content: "";
        width: 100%;
        display: block;
        height: 5px;
        background-color: red;
    }
`;

const StickMan: React.FC = () => {
    return <StickManStyled />;
};

export default StickMan;
