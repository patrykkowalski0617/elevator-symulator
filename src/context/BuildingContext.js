import React, { createContext, useState } from "react";

export const BuildingContext = createContext();

const BuildingContextProvider = props => {
    const [numberOfFloors, setNumberOfFloors] = useState(8);

    return (
        <BuildingContext.Provider
            value={{
                numberOfFloors,
                setNumberOfFloors
            }}
        >
            {props.children}
        </BuildingContext.Provider>
    );
};

export default BuildingContextProvider;
