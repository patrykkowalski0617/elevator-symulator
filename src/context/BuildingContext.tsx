import React, { createContext, useState } from "react";

interface IContextProps {
    numberOfFloors: number;
    setNumberOfFloors: (n: number) => void;
    numberOfCars: number;
    setNnmberOfCars: (n: number) => void;
    automotion: boolean;
    setAutomotion: (b: boolean) => void;
    creatingStickMan: boolean;
    setCreatingStickMan: (arr: boolean) => void;
}

export const BuildingContext = createContext({} as IContextProps);

const BuildingContextProvider = (props: { children: React.ReactNode }) => {
    const [numberOfFloors, setNumberOfFloors] = useState(11);
    const howManyCars = Math.floor(numberOfFloors / 5);
    const [numberOfCars, setNnmberOfCars] = useState(howManyCars);
    const [automotion, setAutomotion] = useState(false);
    const [creatingStickMan, setCreatingStickMan] = useState<boolean>(false);

    return (
        <BuildingContext.Provider
            value={{
                numberOfFloors,
                setNumberOfFloors,
                numberOfCars,
                setNnmberOfCars,
                automotion,
                setAutomotion,
                creatingStickMan,
                setCreatingStickMan
            }}
        >
            {props.children}
        </BuildingContext.Provider>
    );
};

export default BuildingContextProvider;
