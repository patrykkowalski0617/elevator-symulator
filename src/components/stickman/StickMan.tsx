import React, { useContext } from "react";
import { BuildingContext } from "../../context/BuildingContext";
import { floorColor } from "../../style_mixin";
import { Container, StickManStyled } from "./StickManStyled";

type StickManProps = {
    stickId: number;
    destination: number;
    getIn: boolean;
    numberOfPassengers: number;
    assignedCar: number | null;
    place: string;
};

const StickMan = ({
    stickId,
    destination,
    getIn,
    numberOfPassengers,
    assignedCar,
    place
}: StickManProps) => {
    const { numberOfFloors, carWidth } = useContext(BuildingContext);

    const color = floorColor({ numberOfFloors, floorNumber: destination });

    return (
        <Container
            stickId={stickId}
            getIn={getIn}
            numberOfPassengers={numberOfPassengers}
            assignedCar={assignedCar}
            carWidth={carWidth}
            place={place}
        >
            <p>{destination}</p>
            <StickManStyled color={color}></StickManStyled>
        </Container>
    );
};

export default StickMan;
