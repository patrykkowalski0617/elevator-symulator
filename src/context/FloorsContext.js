import React, { createContext, useState, useContext } from "react";
import { BuildingContext } from "./BuildingContext";

export const FloorsContext = createContext();

const FloorsContextProvider = props => {
    const { numberOfFloors } = useContext(BuildingContext);

    const [floorsWaitingForCar, setFloorsWaitingForCar] = useState(
        Array(numberOfFloors).fill(false)
    );

    const addFloorWaitingForCar = floorNumber => {
        const _floorsWaitingForCar = [...floorsWaitingForCar];
        _floorsWaitingForCar.splice(floorNumber, 1, true);
        setFloorsWaitingForCar([..._floorsWaitingForCar]);
    };

    return (
        <FloorsContext.Provider
            value={{
                floorsWaitingForCar,
                addFloorWaitingForCar,
                setFloorsWaitingForCar
            }}
        >
            {props.children}
        </FloorsContext.Provider>
    );
};

export default FloorsContextProvider;
