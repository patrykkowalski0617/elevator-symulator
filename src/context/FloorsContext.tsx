import React, { createContext, useState, useContext } from "react";
import { BuildingContext } from "./BuildingContext";

interface IContextProps {
    floorsWaitingForCar: boolean[];
    addFloorWaitingForCar: (floorNumber: number) => void;
    setFloorsWaitingForCar: (arr: boolean[]) => void;
    allStickMansDestinations: number[][];
    addAllStickMansDestinations: (
        floorNumber: number,
        howMany: number,
        destination: number
    ) => void;
}

export const FloorsContext = createContext({} as IContextProps);

const FloorsContextProvider = (props: { children: React.ReactNode }) => {
    const { numberOfFloors } = useContext(BuildingContext);

    const [floorsWaitingForCar, setFloorsWaitingForCar] = useState(
        Array(numberOfFloors).fill(false)
    );

    const [allStickMansDestinations, setallStickMansDestinations] = useState<
        number[][]
    >(Array(numberOfFloors).fill([]));

    const addFloorWaitingForCar = (floorNumber: number) => {
        const _floorsWaitingForCar = [...floorsWaitingForCar];
        _floorsWaitingForCar.splice(floorNumber, 1, true);
        setFloorsWaitingForCar([..._floorsWaitingForCar]);
    };

    const addAllStickMansDestinations = (
        floorNumber: number,
        howMany: number,
        destination: number
    ) => {
        const _allStickMansDestinations: number[][] = [
            ...allStickMansDestinations
        ];
        const newStickMans: number[] = [];
        // push number of destinations for each new StickMan
        for (let i = 0; i < howMany; i++) {
            newStickMans.push(destination);
        }

        const replaceStickMans = newStickMans.concat(
            _allStickMansDestinations[floorNumber]
        );
        _allStickMansDestinations.splice(floorNumber, 1, replaceStickMans);

        setallStickMansDestinations([..._allStickMansDestinations]);
    };

    return (
        <FloorsContext.Provider
            value={{
                floorsWaitingForCar,
                addFloorWaitingForCar,
                setFloorsWaitingForCar,
                allStickMansDestinations,
                addAllStickMansDestinations
            }}
        >
            {props.children}
        </FloorsContext.Provider>
    );
};

export default FloorsContextProvider;
