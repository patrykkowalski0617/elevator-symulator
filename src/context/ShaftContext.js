import React, { createContext, useState, useContext } from "react";
import { BuildingContext } from "./BuildingContext";

export const ShaftContext = createContext();

const ShaftContextProvider = props => {
    const { numberOfCars } = useContext(BuildingContext);
    const initArr = val => Array(numberOfCars).fill(val);

    const [carCurrenFloorCon, setCarCurrenFloorCon] = useState(initArr(0)); // next floor on way or floor that has been just reached, but not floor which has been just left

    // console.log(carCurrenFloorCon);

    return (
        <ShaftContext.Provider
            value={{
                carCurrenFloorCon,
                setCarCurrenFloorCon
            }}
        >
            {props.children}
        </ShaftContext.Provider>
    );
};

export default ShaftContextProvider;
