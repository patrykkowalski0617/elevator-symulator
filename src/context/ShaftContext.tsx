import React, { createContext, useState, useContext, useEffect } from "react";
import { BuildingContext } from "./BuildingContext";

interface IContextProps {
    allCarsCurrentFloor: number[];
    updateCarCurrentFloor: (carId: number, currentFloor: number) => void;
    allCarsFloorAssignments: number[][];
    addCarFloorAssignment: (carId: number, floorNumber: number) => void;
    removeCarFloorAssignment: (carId: number, floorNumber: number) => void;
    allCarsState: string[];
    updateCarState: (carId: number, state: string) => void;
    allCarsStickMans: {
        lifeState: string;
        destination: number;
        carId: number | null;
        placeInCar: number | null;
    }[][];
    addPassengers: (
        data: {
            destination: number;
            direction: string | null;
            lifeState: string;
            carId: number | null;
            placeInCar: number | null;
        }[]
    ) => void;
    allCarsDirection: string | null[];
    updateCarDirection: (carId: number, direction: string) => void;
}

export const ShaftContext = createContext({} as IContextProps);

const ShaftContextProvider = (props: { children: React.ReactNode }) => {
    const { numberOfCars } = useContext(BuildingContext);
    const initArr = (val: any) => Array(numberOfCars).fill(val);

    // next floor on way or floor that has been just reached, but not floor which has been just left
    const [allCarsCurrentFloor, setAllCarsCurrentFloor] = useState(initArr(0));
    // ready, go-up, go-down, go-up-door-open, go-down-door-open
    const [allCarsState, setAllCarsState] = useState(initArr("ready"));
    const [allCarsFloorAssignments, setAllCarsFloorAssignments] = useState(
        initArr([])
    );
    const [allCarsStickMans, setAllCarsStickMans] = useState(initArr([]));
    const [allCarsDirection, setAllCarsDirection] = useState(initArr(null));

    const updateCarCurrentFloor = (carId: number, currentFloor: number) => {
        const _allCarsCurrentFloor = [...allCarsCurrentFloor];
        _allCarsCurrentFloor.splice(carId, 1, currentFloor);
        setAllCarsCurrentFloor([..._allCarsCurrentFloor]);
    };

    const updateCarState = (carId: number, state: string) => {
        const _allCarsState = [...allCarsState];
        _allCarsState.splice(carId, 1, state);
        setAllCarsState([..._allCarsState]);
    };

    const addCarFloorAssignment = (carId: number, floorNumber: number) => {
        if (carId !== null) {
            const _allCarsFloorAssignments = [...allCarsFloorAssignments];
            const carFloorAssignment = [..._allCarsFloorAssignments[carId]];
            const carFloorAssignmentUpdated = [
                ...carFloorAssignment,
                floorNumber
            ];
            _allCarsFloorAssignments.splice(
                carId,
                1,
                carFloorAssignmentUpdated
            );
            setAllCarsFloorAssignments([..._allCarsFloorAssignments]);
        }
    };

    const [_allCarsStickMans, set_allCarsStickMans] = useState(
        allCarsStickMans
    );
    useEffect(() => {
        set_allCarsStickMans(allCarsStickMans);
    }, [allCarsStickMans]);
    const addPassengers = (
        data: {
            destination: number;
            direction: string | null;
            lifeState: string;
            carId: number | null;
            placeInCar: number | null;
        }[]
    ) => {
        if (data[0]) {
            const carId =
                typeof data[0].carId === "number" ? data[0].carId : -1;
            if (allCarsStickMans[carId].length < 4) {
                const carPassengers = [..._allCarsStickMans[carId], ...data];
                _allCarsStickMans.splice(carId, 1, carPassengers);
                setAllCarsStickMans([..._allCarsStickMans]);
            } else {
                console.warn("car is full!");
            }
        } else {
            console.warn(
                "there are no stickmen on this floor who have the same direction as the car!"
            );
        }
    };

    const removeCarFloorAssignment = (carId: number, floorNumber: number) => {
        const _allCarsFloorAssignments = [...allCarsFloorAssignments];
        const carFloorAssignment = [..._allCarsFloorAssignments[carId]];
        const index = carFloorAssignment.indexOf(floorNumber);
        carFloorAssignment.splice(index, 1);
        const carFloorAssignmentUpdated = [...carFloorAssignment];
        _allCarsFloorAssignments.splice(carId, 1, carFloorAssignmentUpdated);
        setAllCarsFloorAssignments([..._allCarsFloorAssignments]);
    };

    const updateCarDirection = (carId: number, direction: string) => {
        const _allCarsDirection = allCarsDirection;
        _allCarsDirection.splice(carId, 1, direction);
        setAllCarsDirection([..._allCarsDirection]);
    };

    return (
        <ShaftContext.Provider
            value={{
                allCarsCurrentFloor,
                updateCarCurrentFloor,
                allCarsFloorAssignments,
                addCarFloorAssignment,
                removeCarFloorAssignment,
                allCarsState,
                updateCarState,
                allCarsStickMans,
                addPassengers,
                allCarsDirection,
                updateCarDirection
            }}
        >
            {props.children}
        </ShaftContext.Provider>
    );
};

export default ShaftContextProvider;
