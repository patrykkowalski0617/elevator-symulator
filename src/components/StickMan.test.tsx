import React from "react";
import StickMan from "./StickMan";
import renderer from "react-test-renderer";
import ShaftContext from "../context/ShaftContext";
import BuildingContext from "../context/BuildingContext";
import FloorsContext from "../context/FloorsContext";

test("StickMan - render and match snapshot", () => {
    const tree = renderer.create(
        <BuildingContext>
            <ShaftContext>
                <FloorsContext>
                    <StickMan></StickMan>
                </FloorsContext>
            </ShaftContext>
        </BuildingContext>
    );
    expect(tree.toJSON()).toMatchSnapshot();
});
