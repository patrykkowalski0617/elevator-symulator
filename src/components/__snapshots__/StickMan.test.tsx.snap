// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StickMan - render and match snapshot 1`] = `
<div
  className="sc-AxjAm hvJXFA"
/>
`;
